libsolidity/codegen/ArrayUtils.cpp

```diff
@@ -818,30 +818,25 @@ void ArrayUtils::incrementDynamicArraySize(ArrayType const& _type) const
 		// lowest-order byte (we actually use a mask with fewer bits) must
 		// be (31*2+0) = 62
 
+		m_context << Instruction::DUP1 << Instruction::SLOAD << Instruction::DUP1;
+		m_context.callYulFunction(m_context.utilFunctions().extractByteArrayLengthFunction(), 1, 1);
 		m_context.appendInlineAssembly(R"({
-			let data := sload(ref)
-			let shifted_length := and(data, 63)
 			// We have to copy if length is exactly 31, because that marks
 			// the transition between in-place and out-of-place storage.
-			switch shifted_length
-			case 62
+			switch length
+			case 31
 			{
 				mstore(0, ref)
 				let data_area := keccak256(0, 0x20)
 				sstore(data_area, and(data, not(0xff)))
-				// New length is 32, encoded as (32 * 2 + 1)
-				sstore(ref, 65)
-				// Replace ref variable by new length
-				ref := 32
+				// Set old length in new format (31 * 2 + 1)
+				ref := 63
 			}
-			default
-			{
-				sstore(ref, add(data, 2))
-				// Replace ref variable by new length
-				if iszero(and(data, 1)) { data := shifted_length }
-				ref := add(div(data, 2), 1)
-			}
-		})", {"ref"});
+			sstore(ref, add(data, 2))
+			// return new length in ref
+			ref := add(length, 1)
```

Comment:
I think the `ref := 63` part should not be here. This assignment puts the return value in it but you still need its original value or `sstore(ref, add(data, 2))` will write to the wrong storage slot.

Reply:
but we also have to change the length encoding...

Reply:
I think the fix is to use `data := 63`